added negative constraints

mjpost opened the pull request:
This implements negative constraints in beam search, i.e., phrases that are not permitted to appear in the output. It works by tracking the progress of each negative constraint; if the last word is reached, the corresponding entry in the `scores` array is set to `np.inf` prior to the call to `topk()`.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [ ] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [ ] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

davvil commented on sockeye/inference.py:
```diff
@@ -691,12 +691,22 @@ def make_input_from_json_string(sentence_id: int, json_string: str) -> Translato
+
+        # List of phrases that must appear in the output
         constraints = jobj.get(C.JSON_CONSTRAINTS_KEY)
         if isinstance(constraints, list) and len(constraints) > 0:
```
Why the `isinstance`? Wouldn't it just work with `if constraints:`? Then you catch the case where it is `None` and len() == 0 in one go.

mjpost replied:
Hmm, good point.

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -441,12 +542,20 @@ def main():
 
+    If you pass `--avoid` to the script, the constraints will be generated as negative constraints, instead:
+
+        echo -e "Dies ist ein Test .\tThis is\ttest" | python3 -m sockeye.lexical_constraints --avoid
```
Ask the german guys, but I think "Das ist ein Test" sounds better.

mjpost replied:
Not German but agree.

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -25,6 +25,104 @@
+class AvoidPhrase:
+    """
+    Represents a negative phrasal constraint. Tracks progress through a phrase
+    and returns a word ID if that word ID would complete the phrase.
```
Returns the word ID in the `avoid()` method.

mjpost replied:
Yes, clearer.

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -25,6 +25,104 @@
+        :param word_id: The word that was just generated.
+        """
+        if self.phrase[self.last_consumed + 1] == word_id:
+            return AvoidPhrase(self.phrase, self.last_consumed + 1)
```
Do you need to return a new object? Couldn't you just update your last_consumed index as in the else: case? If so, you wouldn't actually need to return anything and the loop in AvoidBatch would be more efficient.

mjpost replied:
AvoidPhrase objects can be shared so long as they have the same state. This is accomplished by returning a new object every time the state advances. This way every hypothesis has its own record of its progress through each phrase.

Put another way, the AvoidPhrase object is a reference shared across hypotheses in the beam. When any hypothesis advances, it gets a new version of the object that is specific to itself.

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -25,6 +25,104 @@
+
+    def avoid(self) -> int:
+        """
+        Returns an integer ID that should be avoided.
```
I would write "word ID".

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -25,6 +25,104 @@
+        for i, items in enumerate(self.avoid_list):
+            for phrase in items:
+                word_id = phrase.avoid()
+                if word_id:
```
Not strictly necessary, but I would be explicit here and put `if word_id > 0:', as it cannot be None.

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -455,14 +564,30 @@ def main():
+            else:
+                constraints.append(item)
+
+        if len(constraints):
```
`if constraints:`?

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -455,14 +564,30 @@ def main():
+
+        if len(constraints):
             obj['constraints'] = constraints
+        if len(avoid_list):
```
`if avoid_list:`?

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -25,6 +25,106 @@
+        if self.phrase[self.last_consumed + 1] == word_id:
+            return AvoidPhrase(self.phrase, self.last_consumed + 1)
+        elif self.last_consumed != -1:
+            return AvoidPhrase(self.phrase, self.last_consumed)
```
self.last_consumed or -1?

mjpost replied:
good catch.

davvil commented:
It would be useful to have a test for this. Would it be easy to adapt the ones for constrained decoding?

mjpost commented:
agree on the need for a test, will write one.

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -24,106 +24,139 @@
 # Represents a list of raw constraints for a sentence. Each constraint is a list of target-word IDs.
 RawConstraintList = List[List[int]]
 
+class AvoidTrie:
```
Hmm, it is a bit of a strange Trie. The nodes at level i contains a set of phrases the of length i+1 (with root being level 0), instead of the implicit representation of the path up to the node. Is this because of some optimization?

 **Update:** Probably this is to support the avoid() function below, right?

mjpost replied:
Yes, to support `avoid()`. At each trie node, we record all words that should not be generated from that state. A more principled trie would generate arcs leading to terminal states, but this is more efficient for our purposes. Arcs are only generated if they lead to internal states of a phrase that must be avoided.

davvil commented on sockeye/lexical_constraints.py:
```diff
@@ -24,106 +24,139 @@
 
-class AvoidPhrase:
+
+class AvoidState:
```
Do you need a separate class? Can't you just store AvoidTrie (perhaps renamed as AvoidTrieNode for clarity) in AvoidBatch?

mjpost replied:
The state object tracks a position in the trie for the current hypothesis. We could just keep this position, but there are a number of associated operations that make sense to bundle together in a class.

davvil replied:
Yes, the question is if you can't bundle these operations in the trie and avoid yet another wrapper class. As you are already using a non-general, partly non-standard trie I think this would help clarity. But up to you, it's not a big issue.

mjpost replied:
It would be simpler to have one fewer class. I'll give it some thought this evening.

mjpost commented:
Okay, this is now updated to allow both a global block list (`--avoid-list /path/to/list`) plus sentence-specific ones (via the JSON `avoid` keyword). I wrote test cases and rebased on master. It is ready to merge pending CR.

fhieber commented on sockeye/inference.py:
```diff
@@ -1441,6 +1473,12 @@ def _beam_search(self,
+            if self.global_avoid_trie or any(raw_avoid_list):
+                indices = mx.nd.array(avoid_states.avoid(), ctx=self.context)
+                if any(indices):
+                    scores.reshape((self.batch_size * self.beam_size * scores.shape[-1]))[indices] = np.inf
```
This looks strange. Is `scores` after this line reshaped or not?

mjpost replied:
It isn't reshaped. Here is the problem: say word ID 74 is always blocked, batch_size is 1, and beam_size is 5, and the maximum vocabulary size is 10,000. We need to block ID 74 in every row of `scores` (by setting it to `np.inf`). The easiest way to do this is to reshape `scores` into a single-dimension array, and block ID 74, ID 10074, 20074, 30074, and 40074. That's what this code does.

fhieber replied:
I understand what the code does, but was just wondering about the clarity of this expression (using fluent reshape and inplace setting of values).

mjpost replied:
I see. 2d indices in the first place would make more sense, if MXNet supports that.

fhieber commented on sockeye/inference.py:
```diff
@@ -974,8 +987,9 @@ def __init__(self,
         assert C.PAD_ID == 0, "pad id should be 0"
         self.stop_ids = {self.vocab_target[C.EOS_SYMBOL], C.PAD_ID}  # type: Set[int]
         self.strip_ids = self.stop_ids.copy()  # ids to strip from the output
+        UNK_ID = self.vocab_target[C.UNK_SYMBOL]
```
This is not a global variable, so it shouldn't be capitalized.

fhieber commented on sockeye/inference.py:
```diff
@@ -1198,7 +1223,14 @@ def _get_inference_input(self,
+            if trans_input.avoid_list is not None:
+                raw_avoid_list[j] = [data_io.tokens2ids(phrase, self.vocab_target) for phrase in
+                                     trans_input.avoid_list]
+                UNK_ID = self.vocab_target[C.UNK_SYMBOL]
```
same here. Why not 'self' it in the Translator constructor, similar to `self.start_ids` etc.

fhieber commented on sockeye/inference.py:
```diff
@@ -1574,6 +1574,8 @@ def _beam_search(self,
             for ms in model_states:
                 ms.sort_state(best_hyp_indices)
 
+#            self._print_beam(sequences, scores_accumulated, finished, inactive, constraints, t)
```
should be removed

fhieber commented on sockeye/inference.py:
```diff
@@ -1441,6 +1482,12 @@ def _beam_search(self,
+            if self.global_avoid_trie or any(raw_avoid_list):
+                indices = avoid_states.avoid()
+                if any(indices):
+                    scores[indices] = np.inf
```
this is much clearer thanks. 
At some point (not scope of this PR of course) we should think about how to move the constrained/avoid logic to proper operators/gluon blocks that are potentially hybridizable. As seen recently, these imperative indexing ops are quite slow.

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -14,7 +14,7 @@
 import logging
 from operator import attrgetter
-from typing import List, Optional, Tuple, Set
+from typing import List, Optional, Tuple, Set, Dict
```
[nit] could be ordered alphabetically :)

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+    def __init__(self,
+                 raw_phrases: Optional[RawConstraintList] = None) -> None:
+        self.final_ids = set()  # type: Set[int]
+        self.children = {}      # type: Dict[int,'AvoidTrie']
```
[nit] just 2 spaces after code for inline comments (pep8 formatting)

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+                self.children[next_word] = AvoidTrie()
+            self.step(next_word).add_phrase(phrase[1:])
+
+    def step(self, word_id: int) -> 'AvoidTrie':
```
`Optional['AvoidTrie']`

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+            for phrase in raw_phrases:
+                self.add_phrase(phrase)
+
+    def __str__(self):
```
-> str

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+        self.local_avoid_states = []  # type: List[AvoidState]
+
+        # Store the global trie for each hypothesis
+        if global_avoid_trie:
```
i'd explicitly check for None here

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+            self.global_avoid_states = [AvoidState(global_avoid_trie)] * batch_size * beam_size
+
+        # Store the sentence-level tries for each item in their portions of the beam
+        if avoid_list:
```
same here. if the list is empty the for loop is not entered

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+                if word_id > 0:
+                    to_avoid.add((i, word_id))
+
+        x_list = []
```
could use some typing annotation

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+
+        :return: Two lists of indices: the x coordinates and y coordinates.
+        """
+        to_avoid = set()
```
could use some typing annotation

fhieber commented on test/unit/test_constraints.py:
```diff
@@ -191,3 +190,98 @@ def test_constraints_init_batch(raw_constraint_lists):
+    # first two hypotheses blocked on 19 (= 19 and 119), next two on 20 (= 220 and 320)
+    (None, [[[19]], [[20]]], 2, 2, '', [(0,19), (1,19), (2,20), (3,20)]),
+    # same, but also add global block list to each row
+    (['74'], [[[19]], [[20]]], 2, 2, '', [(0,19), (0,74), (1,19), (1,74), (2,20), (2,74), (3,20), (3,74)]),
```
pep8 formatting

fhieber commented on sockeye/inference.py:
```diff
@@ -1226,8 +1226,8 @@ def _get_inference_input(self,
                                      trans_input.avoid_list]
-                UNK_ID = self.vocab_target[C.UNK_SYMBOL]
-                if any([UNK_ID in phrase for phrase in raw_avoid_list[j]]):
+                unk_id = self.vocab_target[C.UNK_SYMBOL]
```
self?

mjpost commented:
I'm not sure why these tests are failing. They pass locally and the Travis CI reported errors have nothing to do with my changes.

tdomhan commented:
yeah, something is up with Travis! Master is also failing now. Seems to be unrelated to this and the other PRs.

tdomhan commented:
somehow `default_bucket_key` is a list...

tdomhan commented:
looks like pyyaml suddenly loads tuples as lists (the buckets).

mjpost commented:
Hi folks, I think this is ready to merge if someone can give it a last pass.

tdomhan commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+            for raw_phrases in avoid_list:
+                self.local_avoid_states += [AvoidState(AvoidTrie(raw_phrases))] * beam_size
+
+    def reorder(self, indices: mx.nd.array) -> None:
```
to be consistent with the rest of the code these should probably be ```mx.nd.NDArray``` (also below).

mjpost replied:
Ok.

tdomhan commented on sockeye/lexical_constraints.py:
```diff
@@ -24,6 +24,200 @@
+                if word_id > 0:
+                    to_avoid.add((i, word_id))
+
+        x_list = []  # type: List[int]
```
this could be simplified to ```list(zip(*to_avoid))```, right? (generating a list of tuples)

mjpost replied:
Ah, yes, thank you! Hated this bit of code.

mjpost commented on sockeye/inference.py:
```diff
@@ -1421,6 +1456,11 @@ def _beam_search(self,
         constraints = constrained.init_batch(raw_constraint_list, self.beam_size, self.start_id,
                                              self.vocab_target[C.EOS_SYMBOL])
 
+        if self.global_avoid_trie or any(raw_avoid_list):
```
No, we need to maintain negative constraint if either (a) there are global constraints or (b) constraints for any sentence in the batch.

mjpost commented on sockeye/inference.py:
```diff
@@ -691,12 +692,22 @@ def make_input_from_json_string(sentence_id: int, json_string: str) -> Translato
-        return TranslatorInput(sentence_id=sentence_id, tokens=tokens, factors=factors, constraints=constraints)
+
+        # List of phrases to prevent from occuring in the output
+        avoid_list = jobj.get(C.JSON_AVOID_KEY)
```
I added a check that it's a list.

mjpost commented:
Okay, I've pulled everything in.
